### PR TITLE
[FLINK-22636] Group job specific zNodes under /jobs zNode

### PR DESCRIPTION
**Pull request by tillrohrmann**
In order to better clean up job specific HA services, this commit changes the layout of the
zNode structure so that the JobMaster leader, checkpoints and checkpoint counter is no grouped
below the jobs zNode.

Moreover, this commit groups the leaders of the cluster components (Dispatcher, ResourceManager,
RestServer) under /leader.

The new layout looks like:

```
clusterID -> jobgraphs -> <job-id>
          -> jobs -> <job-id> -> checkpoints -> checkpoint-1
                              -> checkpoint_id_counter -> counter
                              -> leader -> latch
                                        -> connection_info
          -> leader -> dispatcher -> latch
                                  -> connection_info
                    -> resource_manager -> latch
                                        -> connection_info
```

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/util/ZooKeeperUtils.java [hunk 1]
```diff
@@ -493,14 +522,17 @@ public static String generateZookeeperPath(String root, String namespace) {
...
 
         // Ensure that the checkpoints path exists
         client.newNamespaceAwareEnsurePath(path).ensure(client.getZookeeperClient());
 
         // All operations will have the path as root
-        return client.usingNamespace(generateZookeeperPath(client.getNamespace(), path));
+        return client.usingNamespace(
+                // Curator prepends a '/' manually and throws an Exception if the
+                // namespace starts with a '/'.
+                generateZookeeperPath(client.getNamespace(), path).substring(1));
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/util/ZooKeeperUtils.java:535]: maybe add a check that the first character is really a `/`

File: flink-runtime/src/main/java/org/apache/flink/runtime/util/ZooKeeperUtils.java [hunk 2]
```diff
@@ -466,20 +475,40 @@ public static ZooKeeperCheckpointIDCounter createCheckpointIDCounter(
...
+    /** Creates a ZooKeeper path of the form "/root/namespace". */
     public static String generateZookeeperPath(String root, String namespace) {
-        if (!namespace.startsWith("/")) {
-            namespace = '/' + namespace;
-        }
+        final String result =
+                Stream.of(root, namespace)
+                        .map(ZooKeeperUtils::trimSlashes)
+                        .filter(s -> !s.isEmpty())
+                        .collect(Collectors.joining("/"));
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/util/ZooKeeperUtils.java:484]: nit:
```suggestion
                        .collect(Collectors.joining("/", "/", ""));
```

**Review by zentol — CHANGES_REQUESTED** [review]
Funtionality-wise this is fine, but there appears to be an off-by-one error somewhere in the path processing:
```
[zk: 127.0.0.1:2181(CONNECTED) 3] ls /
[flin, zookeeper]
[zk: 127.0.0.1:2181(CONNECTED) 4] ls /flin
[defau, defaul]
[zk: 127.0.0.1:2181(CONNECTED) 5] ls /flink/defaul
Node does not exist: /flink/defaul
[zk: 127.0.0.1:2181(CONNECTED) 6] ls /flink/defau
Node does not exist: /flink/defau
[zk: 127.0.0.1:2181(CONNECTED) 7] ls /flin/defaul
[job, jobgraphs, jobs, leade, running_job_registry]
[zk: 127.0.0.1:2181(CONNECTED) 8]
```

File: flink-runtime/src/main/java/org/apache/flink/runtime/leaderretrieval/ZooKeeperLeaderRetrievalDriver.java [hunk 3]
```diff
@@ -188,6 +190,15 @@ public void unhandledError(String s, Throwable throwable) {
 
     @Override
     public String toString() {
-        return "ZookeeperLeaderRetrievalDriver{" + "retrievalPath='" + retrievalPath + '\'' + '}';
+        return "ZookeeperLeaderRetrievalDriver{"
+                + "retrievalPath='"
```
[the remark this answers is not in the document]
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/leaderretrieval/ZooKeeperLeaderRetrievalDriver.java:194]: Will update it.

File: flink-runtime/src/main/java/org/apache/flink/runtime/util/ZooKeeperUtils.java (hunk 1 above)
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/util/ZooKeeperUtils.java:535]: Good idea. Will do it.

File: flink-runtime/src/main/java/org/apache/flink/runtime/util/ZooKeeperUtils.java (hunk 2 above)
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/util/ZooKeeperUtils.java:484]: Good point. Will change it.

File: flink-runtime/src/main/java/org/apache/flink/runtime/util/ZooKeeperUtils.java (hunk 1 above)
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/util/ZooKeeperUtils.java:535]: I solved this problem a bit differently. I introduced a `trimStartingSlash` function.

**Comment by tillrohrmann** [comment]
Thanks for the review @zentol. You are right. I messed up the `generateZookeeperPath` function which is causing the off-by-one error. I will fix it with a fixup.

**Review by zentol — APPROVED** [review]

**Pull request closed without merging**